build_tools/cmake/iree_check_test.cmake

```diff
@@ -254,33 +250,83 @@ function(iree_check_single_backend_test_suite)
     ${ARGN}
   )
 
+  # For artifacts compiled with the llvm-cpu backend to be compatible with
+  # runtime builds using sanitizers, a few additional flags must be passed.
+
+  set(_BASE_LABELS ${_RULE_LABELS})
+  # ThreadSanitizer (TSan) is an ABI break: when the host code is built with
+  # TSan, the module must be too, otherwise we get crashes calling module code.
+  if(${_RULE_TARGET_BACKEND} STREQUAL "llvm-cpu")
+    list(APPEND _BASE_LABELS "notsan")
+  endif()
+
   foreach(_SRC IN LISTS _RULE_SRCS)
     get_filename_component(_BASE_NAME ${_SRC} NAME)
     set(_TEST_NAME "${_RULE_NAME}_${_BASE_NAME}")
     iree_check_test(
-      NAME
-        ${_TEST_NAME}
-      SRC
-        ${_SRC}
-      TARGET_BACKEND
-        ${_RULE_TARGET_BACKEND}
-      DRIVER
-        ${_RULE_DRIVER}
-      COMPILER_FLAGS
-        ${_RULE_COMPILER_FLAGS}
-      INPUT_TYPE
-        ${_RULE_INPUT_TYPE}
-      RUNNER_ARGS
-        ${_RULE_RUNNER_ARGS}
-      LABELS
-        ${_RULE_LABELS}
-      TARGET_CPU_FEATURES
-        ${_RULE_TARGET_CPU_FEATURES}
-      DEPENDS
-        ${_RULE_DEPENDS}
-      TIMEOUT
-        ${_RULE_TIMEOUT}
+      NAME ${_TEST_NAME}
+      SRC ${_SRC}
+      TARGET_BACKEND ${_RULE_TARGET_BACKEND}
+      DRIVER ${_RULE_DRIVER}
+      COMPILER_FLAGS ${_RULE_COMPILER_FLAGS}
+      INPUT_TYPE ${_RULE_INPUT_TYPE}
+      RUNNER_ARGS ${_RULE_RUNNER_ARGS}
+      LABELS ${_BASE_LABELS}
+      TARGET_CPU_FEATURES ${_RULE_TARGET_CPU_FEATURES}
+      DEPENDS ${_RULE_DEPENDS}
+      TIMEOUT ${_RULE_TIMEOUT}
     )
+
+    # Define tests for AddressSanitizer (ASan) and ThreadSanitizer (TSan).
+    # Normally test suites should do this sort of branching at the leaves rather
+    # than modify the base CMake function directly, but sanitizers are applied
+    # at the build system uniformly, so until we decouple the test suites from
+    # source builds further this felt like a reasonable compromise.
+    #
+    # TODO(scotttodd): could add `CONDITION` options to
+    #   iree_check_test that would allow for tests being defined but marked
+    #   DISABLED, rather than wrap these with all-or-nothing CMake `if(...)`s?
+    if(_RULE_TARGET_BACKEND STREQUAL "llvm-cpu")
+      if(IREE_ENABLE_ASAN)
+        set(_ASAN_COMPILER_FLAGS ${_RULE_COMPILER_FLAGS})
+        list(APPEND _ASAN_COMPILER_FLAGS "--iree-llvmcpu-link-embedded=false")
+        list(APPEND _ASAN_COMPILER_FLAGS "--iree-llvmcpu-sanitize=address")
+        set(_ASAN_LABELS ${_RULE_LABELS})
+        list(APPEND _ASAN_LABELS "notsan")
```

Comment:
Hum, this is in case both `IREE_ENABLE_ASAN` and `IREE_ENABLE_TSAN` are `ON`?  I don't think we support that case (this would result in incompatible or mutually-overriding `-fsanitize` flags anyway). I'd rather drop that and, if you want, enforce exclusivity of these CMake options at the root CMakeLists where they are defined.

Reply:
@ScottTodd now that I think of it, I'd like to request one more change. Using the `notsan` label to avoid running a test that's bound to fail with TSan, relies on the developer thinking of passing the CTest flag to skip tests that have the `notsan` label. It's not automatic, and so all developers who are used to manually invoking CTest (myself included) as opposed to just relying on the `build_tools/cmake/` scripts, are going to get the errors.

The existing tests using `notsan` are a minority and are often tied to some GPU backend, so the TSan-failure status makes sense and one simply avoids running into them by avoiding enabling GPU backends with TSan.

This would now make every CPU test fail with TSan unless explicitly skipping `notsan` tests -- that would be very confusing.

So in this instance, rather than creating a test that's bound to fail with TSan and applying the `notsan` label to it, I would prefer if we simply conditionally didn't create the test at all. I'm not asking a given test/target to have its meaning change based on sanitizer options, i'm just asking one test/target to not exist with IREE_ENABLE_TSAN (just like the `_tsan` targets do not exist *without* it). WDYT?